subiquity/server/controllers/shutdown.py

```diff
@@ -23,42 +23,65 @@
 from subiquitycore.utils import arun_command, run_command
 
 from subiquity.common.apidef import API
+from subiquity.common.types import ShutdownMode
 from subiquity.server.controller import SubiquityController
 from subiquity.server.controllers.install import ApplicationState
 
 log = logging.getLogger("subiquity.controllers.restart")
 
 
-class RebootController(SubiquityController):
+class ShutdownController(SubiquityController):
 
-    endpoint = API.reboot
+    endpoint = API.shutdown
+    autoinstall_key = 'shutdown'
+    autoinstall_schema = {
+        'type': 'string',
+        'enum': ['reboot', 'poweroff']
+    }
 
     def __init__(self, app):
         super().__init__(app)
-        self.user_reboot_event = asyncio.Event()
-        self.rebooting_event = asyncio.Event()
+        self.user_shutdown_event = asyncio.Event()
+        self.server_reboot_event = asyncio.Event()
+        self.shuttingdown_event = asyncio.Event()
+        self.install_complete = False
+        self.mode = ShutdownMode.REBOOT
 
-    async def POST(self):
+    def load_autoinstall_data(self, data):
+        if data == 'reboot':
+            self.mode = ShutdownMode.REBOOT
+        elif data == 'poweroff':
+            self.mode = ShutdownMode.POWER_OFF
+
+    async def POST(self, mode: ShutdownMode, immediate: bool = False):
+        self.mode = mode
         self.app.controllers.Install.stop_uu()
-        self.user_reboot_event.set()
-        await self.rebooting_event.wait()
+        self.user_shutdown_event.set()
+        if immediate:
+            self.server_reboot_event.set()
+        await self.shuttingdown_event.wait()
 
     def interactive(self):
         return self.app.interactive
 
     def start(self):
+        self.app.aio_loop.create_task(self._wait_install())
         self.app.aio_loop.create_task(self._run())
 
-    async def _run(self):
-        Install = self.app.controllers.Install
-        await Install.install_task
+    async def _wait_install(self):
+        await self.app.controllers.Install.install_task
+        self.install_complete = True
```

Comment:
To my eyes, install_complete is redundant with `self.app.state == ApplicationState.DONE` - am I missing something?

Reply:
Probably not, no.